Add repro for #7034

gridbugs opened the pull request:
Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Add repro for #7034

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>
Signed-off-by: Ali Caglayan <alizter@gmail.com>

rgrinberg commented on test/blackbox-tests/test-cases/github7034.t/inner/dune:
```diff
@@ -0,0 +1,3 @@
+(library
+ (public_name inner)
+ (libraries lwt lwt_log))
```
Can we do without the external dependencies?

rgrinberg commented on test/blackbox-tests/test-cases/github7034.t/run.t:
```diff
@@ -0,0 +1,73 @@
+
+But when lang dune is 3.3 or higher the warning becomes an error:
+  $ cp outer/dune-project outer/dune-project.bak
+  $ sed 's/(lang dune 3.2)/(lang dune 3.3)/' outer/dune-project > outer/dune-project.new
```
Instead of using sed and symlinks, perhaps you can use shell functions? E.g.

```
$ function gen-dune-project() {
> cat >$1/dune-project <<EOF
> (lang dune $2)
> EOF
> }
$ gen-dune-project "outer/" "3.3"
```

gridbugs commented:
I added a helper script for generating dune-project files, and added some more comments describing the test. I also changed the test so that it expects the the build to succeed, so the test will fail until the bug is fixed.

rgrinberg commented:
Thanks. That looks a lot better. Some failures remain:

```
+++ b/_build/.sandbox/f7446cd9c6c5f7e80bf8ca5c779cfaa0/default/test/blackbox-tests/test-cases/github7034.t/run.t.corrected
[32](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:33)
@@ -41,6 +41,9 @@ Change the version back to 3.2 so the inner package builds again:
[33](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:34)
 
[34](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:35)
 Copy the source of the inner package into outer/vendored:
[35](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:36)
   $ cp -r inner outer/vendored
[36](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:37)
+  cp: inner/_build/install/default/lib/inner/inner.cmx: No such file or directory
[37](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:38)
+  cp: inner/_build/install/default/lib/inner/inner.cmt: No such file or directory
[38](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:39)
+  [1]
[39](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:40)
 
[40](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:41)
 Building the outer project works when lang dune is 3.2 or below:
[41](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:42)
   $ ./gen-dune-project outer 3.2 > outer/dune-project
[42](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:43)
@@ -59,8 +62,9 @@ But when lang dune is 3.3 or higher the warning becomes an error:
[43](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:44)
   File "vendored/inner/inner.ml", line 6, characters 11-18:
[44](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:45)
   6 | type t = { x : int }
[45](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:46)
                  ^^^^^^^
[46](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:47)
-  Warning 69 [unused-field]: unused record field x.
[47](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:48)
+  Error (warning 69 [unused-field]): unused record field x.
[48](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:49)
   Leaving directory 'outer'
[49](https://github.com/ocaml/dune/actions/runs/4189703693/jobs/7262333164#step:11:50)
+  [1]
```

gridbugs commented:
Regarding the failure:
```
   $ cp -r inner outer/vendored
+  cp: inner/_build/install/default/lib/inner/inner.cmx: No such file or directory
+  cp: inner/_build/install/default/lib/inner/inner.cmt: No such file or directory
```
I just pushed a change that runs `dune clean --root=inner` bufore copying the inner directory into outer/vendored, but it's interesting that I couldn't reproduce this problem on my mac. It's also frustrating that cp would fail in this way. Any idea off-hand why it got confused?

gridbugs commented:
Oh and regarding this failure:
```
   File "vendored/inner/inner.ml", line 6, characters 11-18:
   6 | type t = { x : int }
                  ^^^^^^^
-  Warning 69 [unused-field]: unused record field x.
```
I intended for it to fail. For repro PRs should the test expect the output to be how dune should behave once the bug is fixed, or should I change the expected output so that the tests passes now?

And finally, the last few lines of the failing test on windows are:
```
  opam64/install.sh
  D:\cygwin\bin\bash.exe C:\hostedtoolcache\windows\opam\0.0.0.2\x64\opam64\install.sh --prefix /usr
  the following programs are not installed properly:
   curl
  I can't proceed :(
  Error: The process 'D:\cygwin\bin\bash.exe' failed with exit code 2
```
...which looks like a problem with our CI.

rgrinberg commented:
>  Any idea off-hand why it got confused?

No idea, sorry. For that reason, I avoid copying in the tests however. Sometimes it can't be helped, but it's good to at least try.

> I intended for it to fail. For repro PRs should the test expect the output to be how dune should behave once the bug is fixed, or should I change the expected output so that the tests passes now?

I think we should promote and let the eventual bug fix update the output once more. You could leave a comment to mention that the current output is undesirable, but it seems clear to me already from your description in the test.
